libfoundation/src/foundation-objc.mm

```diff
@@ -383,6 +383,794 @@ MC_DLLEXPORT_DEF uintptr_t MCObjcObjectGetActionProxySelector(void)
+{
+    self = [super initWithHandlerMapping:aHandlerMapping withContext:aContext];
+    if (self == nil)
+        return nil;
+    
+    m_protocol = aProtocol;
+    
+    return self;
+}
+
+- (void)dealloc
+{
+    [super dealloc];
+}
+
+- (NSMethodSignature *)methodSignatureForSelector:(SEL)aSelector
+{
+    // Return the method signature of the protocol method we have an LCB
+    // implementation for. This causes forwardInvocation to be called.
+    objc_method_description t_desc;
+    if (!_get_method_description(m_protocol, aSelector, t_desc))
+        return nullptr;
+    
+    return [NSMethodSignature signatureWithObjCTypes:t_desc.types];
+}
+
+@end
+
+@implementation com_livecode_MCObjcInformalDelegate
+
+- (id)initForProtocol:(MCArrayRef)aProtocol withHandlerMapping:(MCArrayRef)aHandlerMapping withContext:(MCValueRef)aContext
+{
+    self = [super initWithHandlerMapping:aHandlerMapping withContext:aContext];
+    if (self == nil)
+        return nil;
+    
+    m_protocol = MCValueRetain(aProtocol);
+    
+    return self;
+}
+
+- (void)dealloc
+{
+    MCValueRelease(m_protocol);
+    [super dealloc];
+}
+
+- (NSMethodSignature *)methodSignatureForSelector:(SEL)aSelector
+{
+    // Construct the type signature
+    MCAutoStringRef t_selector;
+    if (!MCStringCreateWithCString(sel_getName(aSelector), &t_selector))
+        return nullptr;
+    
+    MCAutoStringRef t_types;
+    if (!_get_informal_protocol_method_types(*t_selector, m_protocol, &t_types))
+        return nullptr;
+    
+    MCAutoStringRefAsCString t_types_cstring;
+    if (!t_types_cstring.Lock(*t_types))
+        return nullptr;
+    
+    return [NSMethodSignature signatureWithObjCTypes:*t_types_cstring];
+}
+
+@end
+
+@implementation com_livecode_MCObjcDelegate
+
+- (id)initWithHandlerMapping:(MCArrayRef)aHandlerMapping withContext:(MCValueRef)aContext
+{
+    self = [super init];
+    if (self == nil)
+        return nil;
+    
+    m_handlers = MCValueRetain(aHandlerMapping);
+    m_context = aContext != nullptr ? MCValueRetain(aContext) : aContext;
+    
+    return self;
+}
+
+- (void)dealloc
+{
+    MCValueRelease(m_handlers);
+    MCValueRelease(m_context);
+    [super dealloc];
+}
+
+-(MCHandlerRef)handlerFromSelector:(SEL)aSelector
+{
+    MCAutoStringRef t_selector;
+    if (!MCStringCreateWithCString(sel_getName(aSelector), &t_selector))
+        return nullptr;
+    
+    MCNewAutoNameRef t_key;
+    if (!MCNameCreate(*t_selector, &t_key))
+        return nullptr;
+    
+    MCValueRef t_handler;
+    if (!MCArrayFetchValue(m_handlers, false, *t_key, t_handler))
+        return nullptr;
+    
+    return static_cast<MCHandlerRef>(t_handler);
+}
+
+-(BOOL)respondsToSelector:(SEL)aSelector
+{
+    // Check the handler mapping to see if this delegate wants to handle
+    // the incoming selector
+    return [self handlerFromSelector:aSelector] != nullptr;
```

Comment:
@livecodeali should this also check if super responds just in case something checks on NSObject method?

Reply:
Yeah, seems reasonable

Reply:
I think you may have missed adding the super call in here

Reply:
Oh, that's annoying! Must have clobbered it

Reply:
Fixed.